### ISSUE TITLE
IO.select fails on MSVC

### ISSUE THREAD

**Issue by cyberarm** [mruby/mruby#5659]
I'm trying to use `IO.select` to monitor sockets; It works on CRuby but crashes on MRuby.

Script
https://github.com/cyberarm/broadcast_forwarder_wireguard/blob/e50413f37070cdaf88795a06d1f4a18c26bc09b0/broadcast_forwarder_wireguard.rb

I added a `printf("WSAGetLastError: %i\n", WSAGetLastError());` inside of https://github.com/mruby/mruby/blob/1508da474090fd921c9e533ca5c080edc6ff1615/mrbgems/mruby-io/src/io.c#L1261

According to `WSAGetLastError()` its caused by `WSAEINVAL`
https://docs.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2

Backtrace
```
WSAGetLastError: 10022
trace (most recent call last):
        [4] broadcast_forwarder_wireguard.rb:74
        [3] broadcast_forwarder_wireguard.rb:74:in new
        [2] broadcast_forwarder_wireguard.rb:15:in initialize
        [1] broadcast_forwarder_wireguard.rb:33:in monitor_broadcasts
broadcast_forwarder_wireguard.rb:33:in select: Unknown error: 0 - select failed (SystemCallError)
```

Build Config
```ruby
MRuby::Build.new do |conf|
  conf.toolchain
  
  conf.gembox 'default'
  conf.gem :github => 'mattn/mruby-json'
  
  conf.enable_debug
end
```

**Comment by matz** [comment]
Unfortunately, mruby-socket assumes sockets can be represented by file descriptors (as UNIX does), but it's not true for WIN32 system using MSVC. CRuby implements an emulation layer on top of Windows socket. But it's not affordable for us. Your options are:

* use MSYS that provides emulation layer by itself
* use WSL

Our (future) options are:

* leave it as it is
* make mruby-socket to raise a build error on MSVC
* implement emulation layer for mruby (need expert to volunteer)

What's your opinion?

**Comment by cyberarm** [comment]
I see, that is unfortunate; I'll try using MSYS.

In the short term it would be good if it threw a descriptive error on MSVC that it isn't supported.
_Eventually_, it would be nice if it "Just Worked" like CRuby.

**Comment by matz** [comment]
Will do, but I need to find a macro symbol to distinguish MSVC first (`_WIN32` is defined for MSYS too).

**Comment by llothar** [comment]
The ignorance of the Windows platform with the arrogance to call mruby cross platform is unfortunately still high. I tried to raise this point 15 year ago but nobody listened.  

There would be quite a few API changes required. For example symlink on windows depends if it is a symbolic link to a file or a directory and the caller has to know this before invoking symlink, so two new functions are required.  Just one more example. I burned out 10 years ago to talk about it, it's worthless as long as there is no shift in mentality.  mruby could be such a nice embedded language if it would be written with more openess and less unix in mind.

Using MSYS is not a solution.  There are still many things that require to use visual studio compiler and the whole new Windows Application SDK is not going to be able to use gcc/clang despite all the attempts another subgroup at Microsoft made to bring clang to windows.

**Comment by matz** [comment]
I agree with you. We need more contributors with Windows knowledge to be "real" cross-platform.
Besides that, the original Ruby behavior is designed under UNIX mind, we can not change this.
But good point is mruby I/O features are separated in mrbgems, so that we can implement Window I/O mrbgems.

**Comment by Asmod4n** [comment]
If you feel a bit adventurous you could try compiling mruby with https://github.com/jart/cosmopolitan.
It provides a posix API which mruby expects and cross compiles one statically linked binary for 7 different platforms, including windows.

**Comment by matz** [comment]
Following up on this 3 years later — the relevant code has changed substantially, so it's worth re-checking whether the original symptom still reproduces.

**Changes since this issue was filed (2022):**

1. **mruby-io HAL split** (commit 9965f11cfe). `mrb_hal_io_select`, `mrb_hal_io_fdset_*` now live in `mrbgems/mruby-io/ports/win/io_hal.c` with a Winsock-native implementation. The 2022-era code path that mixed POSIX FD assumptions with Winsock no longer exists in master.

2. **Winsock error translation in mruby-socket** (commit f267925d45, 2026). When a Winsock call fails, the underlying `WSAGetLastError()` is now mapped to a POSIX errno before `mrb_sys_fail` runs. So if `IO.select` does still fail on MSVC, the error message would be something useful like `Errno::EINVAL: Invalid argument - select failed` rather than the original "Unknown error: 0".

3. **`_MSC_VER` is the right macro** to distinguish MSVC from MSYS/MinGW (mentioned at the time as a blocker). It's defined only by Microsoft Visual C++ compilers; MSYS/MinGW define `_WIN32` but not `_MSC_VER`. So a descriptive build-time `#pragma message` or a runtime check is straightforward to add when needed.

**What I'd like before deciding next steps:**

Could anyone with an MSVC mruby build try the original repro (https://github.com/cyberarm/broadcast_forwarder_wireguard/blob/e50413f37070cdaf88795a06d1f4a18c26bc09b0/broadcast_forwarder_wireguard.rb) on current master and report back?

Possible outcomes:
- **Works now**: close as resolved by the HAL split.
- **Still fails, but with a sensible Errno**: the underlying limitation persists, but the error is now diagnosable. Add a `_MSC_VER`-gated diagnostic and document.
- **Still fails with a confusing error**: the HAL change didn't reach the actual select path; investigate further.

Without an MSVC environment to test from this side, I can't move the issue forward unilaterally. Leaving open for a Windows tester.

**Comment by Asmod4n** [comment]
@matz can take a look in a few hours.

**Comment by Asmod4n** [comment]
@matz select works now, but its calling a Socket function mruby doesnt have afterwards.

```
.\mruby\bin\mruby.bat .\broadcast_forwarder_wireguard.rb
trace (most recent call last):
        [4] .\broadcast_forwarder_wireguard.rb:74
        [3] .\broadcast_forwarder_wireguard.rb:15:in initialize
        [2] .\broadcast_forwarder_wireguard.rb:37:in monitor_broadcasts
        [1] .\broadcast_forwarder_wireguard.rb:38:in monitor_broadcasts
.\broadcast_forwarder_wireguard.rb:44:in forward_broadcast_to_interfaces: undefined method 'ip_address_list' for Class (NoMethodError)
```

guess this can be closed now @cyberarm

**Comment by matz** [comment]
Thanks @Asmod4n for the verification!

The original symptom — `IO.select` failing with `WSAEINVAL`/"Unknown error: 0" on MSVC — is **resolved**. The HAL split (`9965f11cfe`) gave Windows its own Winsock-native `select` path, which is what made cyberarm's repro proceed past line 38.

The new error you hit (`undefined method 'ip_address_list' for Class`) is a separate concern: `Socket.ip_address_list` is a method that exists in CRuby's stdlib socket but isn't implemented in mruby-socket. I'll address that as a follow-up.

Closing this one as resolved by the HAL split.
